### ISSUE TITLE
Issue in embedding image

### ISSUE THREAD

**Issue by PankajRasgania** [turnermm/ckgedit#543]
When using DokuWiki file browser in ckgedit for embedding images, the ckgedit popup doesn't fill url and other details, if the image is selected after a search. It works fine if the image in dokuwiki file browser is selected directly without any search.

**Comment by turnermm** [comment]
Either I can't reproduce this or I don't follow this explanation.

**Comment by PankajRasgania** [comment]
See the following images:
![1](https://user-images.githubusercontent.com/19621227/81294683-dc0ecc80-908c-11ea-9432-7678b5b3ed73.PNG)
Searching for an image in my media library
![2](https://user-images.githubusercontent.com/19621227/81294703-e16c1700-908c-11ea-97cc-b38faf0c278d.PNG)
Selecting the image
![3](https://user-images.githubusercontent.com/19621227/81294721-e7fa8e80-908c-11ea-89b5-fb042d926238.PNG)
After selecting the image, url field of ckgedit popup remains empty
![4](https://user-images.githubusercontent.com/19621227/81294739-ecbf4280-908c-11ea-84ac-676908e9c9c1.PNG)
The above images are from a fresh dokuwiki install with no plugins/template installed, except ckgedit

**Comment by turnermm** [comment]
I've temporarily removed the search function from the media window.  I'm currently working on update to the handling of cookies required by current browsers and will try to get back to this when I have the time.

**Comment by PankajRasgania** [comment]
Okay.

**Comment by turnermm** [comment]
I tried to work this out, put a lot of time in, but couldn't fix it.  Sorry.

**Comment by turnermm** [comment]
Persistence paid off.  I figured this out:
  https://github.com/turnermm/ckgedit/archive/medmgr_sel.zip
It's not merged into the master yet. Could try the above and report back?
Thanks.

**Comment by turnermm** [comment]
I hadn't tested this in ubuntu and there unfortunately it fails.  It works in Centos and Windows.  I'll see what I can do.

**Comment by PankajRasgania** [comment]
I updated the ckgedit plugin after your commit. Unfortunately, it is still not working for me. I checked it on two dokuwiki installations. One of it doesn't have any other plugin.

**Comment by turnermm** [comment]
I initially thought it wasn't working as well.  Which is what accounts for my earlier response.  As you can see there is a thumbs up emoji tere, so another user has found that it works.  Try deleting your dokuwiki cache (all directories in data/cache), and refresh your browser cache.  And make sure that ckgedit/scripts/mediamgr.js has this correction:
https://github.com/turnermm/ckgedit/commit/0230ca04a077714fe9ff63cd58b2d1ee442fe647#diff-e4285cf22040e33ed07ee062af8f2a84
If it doesn't work, let me know the details of you OS and your dokuwiki.

**Comment by PankajRasgania** [comment]
I checked it again. The commit is reflected in the scripts/mediamgr.js as seen from developer tools in the browser. But it is still not working for me. However, the URL gets load in CKG filebrowser but the format is wrong. It shows "/dokuwiki/data/media/{filename}" in the URL. Whereas while using DW filebrowser no field of the form is filled.

I check it on two installations. One is running on windows 10 (using XAMPP) (no plugins installed except ckgedit), and the other is running on Hostgator Shared Hosting (CentOS Linux x86). Both installations are having this issue. I checked after clearing dokuwiki cache and browser cache, still no success. I checked it on a new browser too without success. 

Dokuwiki version is 2018-04-22b "Greebo".

**Comment by turnermm** [comment]
This appears to be your problem:
> However, the URL gets load in CKG filebrowser but the format is wrong. It shows "/dokuwiki/data/media/{filename}

After you have clicked insert in the DokuWiki image dialog, you should get this in the ckgeditor's image dialog:
![2020-05-16_09h21_35](https://user-images.githubusercontent.com/176447/82122369-df731980-9758-11ea-9727-f13b6ab4a6fe.png)

Notice, the url uses Dokuwiki's `fetch` . It is the same for each of the possible settings url types: htaccess, dokuwiki internal, and none.

**Comment by PankajRasgania** [comment]
When using native DokuWiki filebrowser, the none of the form field gets filled. It's the same as in the pics earlier. I don't what I am missing, but it doesn't seem to work.

**Comment by turnermm** [comment]
Can you give me an update on this?

**Comment by PankajRasgania** [comment]
Yes this issue still exists when using search in Dokuwiki Filebrowser. 
In CKG Filebrowser, the preview is not shown in the dialog box and editor, but the image is embedded when saved.

**Comment by PankajRasgania** [comment]
I noticed something which might be the cause of the problems.
When CKGEdit opens media manager to embed image it has some GET parameters `?ns=&edid=wiki__text&onselect=ckg_edit_mediaman_insert&ckg_media=img&CKEditor=wiki__text&CKEditorFuncNum=1&langCode=en` which I think serve the purpose of associating user action to a corresponding function.

But when the search results in dokuwiki filebrowser are rendered there are no GET parameters in the url. May be the absence of these parameters is causing the problem..

**Comment by turnermm** [comment]
Inserting the link was the purpose of https://github.com/turnermm/ckgedit/commit/0230ca04a077714fe9ff63cd58b2d1ee442fe647#diff-e4285cf22040e33ed07ee062af8f2a84

But for some reason it is being ignored in some browsers.  When I get the time, I'll take another look at it.  I still have to go through my plugins and make sure they are compatible with hogfather.  And there are 30 ot 40 of them, so it will take some time.

**Comment by PankajRasgania** [comment]
Sure. It's not a a big issue for me either as if I browse to the image without using search, then it works perfectly. Take your time.

**Comment by PankajRasgania** [comment]
Just found a quick fix. 
The search form in file browser popup with id `#dw__mediasearch` has action attribute set as `/dokuwiki/lib/exe/mediamanager.php`. The script just needs to change this attribute to `/dokuwiki/lib/exe/mediamanager.php?ns=&edid=wiki__text&onselect=ckg_edit_mediaman_insert&ckg_media=img&CKEditor=wiki__text&CKEditorFuncNum=1&langCode=en`


Then it would work for all cases.

The following code works fine.

```
document.querySelectorAll("#dw__mediasearch").forEach(function(form){
form.setAttribute('action', '/dokuwiki/lib/exe/mediamanager.php?ns=&edid=wiki__text&onselect=ckg_edit_mediaman_insert&ckg_media=img&CKEditor=wiki__text&CKEditorFuncNum=1&langCode=en');});
```

**Comment by turnermm** [comment]
That is a very interesting suggestion.  But my preference would be to make the fix local and a fix that can be packaged with ckgedit, so I revised  ckgedit/scripts/mediamgr.js .  I'm not sure it's the final jQuery formulation; I'm not a master of jQuery.  Attached is a copy of the scripts/mediamgr.js file.  See if works for you.

**Comment by turnermm** [comment]
[mediamgr.zip](https://github.com/turnermm/ckgedit/files/4876649/mediamgr.zip)

Sorry.  I forgot to attach the file. I'm eager to see how your tests work out.

**Comment by PankajRasgania** [comment]
Yes, my intention too was to add the code in mediamgr.js.

Yes, your commit in `scripts/mediamgr.js` is working good. There is just one small issue though, when we switch namespace in filebrowser, the contents (including the form we want to change) gets loaded using Ajax. So in that case the script is unable to change the new dynamically injected search form. 

I have made some changes to fix this.
[mediamgr.zip](https://github.com/turnermm/ckgedit/files/4877215/mediamgr.zip)

Summary of changes:

1. Removed the condition check `if(!jQuery("mediamanager__uploader").length)` as the change in action attribute is pretty harmless on its own and should occur everytime.
2. Add the code for changing action attribute in `jQuery(document).ajaxComplete` function, so that the new forms that are added using Ajax also have the desired action url.

See if this works and meets all your requirements.

**Comment by turnermm** [comment]
Thank you.  As far as I can tell it's working very well and I was very happy to have your help.  I've merged it into the stable-master, together with the fix for the end-of-line new line.  If there's anything else among the still open issues that you think you can help with, please look at it.

**Comment by PankajRasgania** [comment]
I would be happy to help where-ever I can. I again want to thank you for all the efforts you are putting in maintaining this plugin.

**Comment by turnermm** [comment]
@PankajRasgania I hadn't gotten back to you sooner today because I found a problem with the media links which I finally traced back to some changes I made to the CKEdiitor link tool.  These changes had only to do with external links, but somehow they affected media links. (The law of unintended consequences.)  You may not have downloaded a recent `stable-master` and so it would not have shown up for you.  Before I check out why this happened, I'd like to get the things we are working on finished.  I'll get to that tomorrow.

**Comment by turnermm** [comment]
I see this morning that you've been working on the hogfather branch.  I'll wait to see what t he outcome is. In the meantime, I've managed to fix the stable-master branch, and all is working.  But I had to  revert back to an earlier version of the link tool.  The stable-master is important because that is currently the download version, and the one which most sites still use.  Hogfather is treated as a special case, at lease until I manage to get a version which works on both stable-master and hogfather.

**Comment by PankajRasgania** [comment]
```
/* update search form action */
                 var valu = jQuery("#dw__mediasearch  input[type='text").val();
                 if(valu == ""){
                     var url = jQuery("form#dw__mediasearch").attr('action'); 
                      jQuery("form#dw__mediasearch").attr('action', url+_action); 
                  };
```
This piece of code actually results in undesired behaviour. It replaces the action attribute only when the search input is empty (which it is for the first time). But lets say I searched for something. Now the resulting search results page will have not have search input empty. In this case, the action attribute will not be updated. So if I perform another search from this page, the link wont be inserted as there are no GET parameter in it.

The `mediamgr.js` that I finally uploaded in #565 seems to be working. So no extra condition check would be required unless we see undesired behaviour. Also as I said earlier, adding GET parameters is harmless and should be added to all forms other wise the plugin will break.

**Comment by turnermm** [comment]
>  But lets say I searched for something. Now the resulting search results page will have not have search input empty.

Yes, that makes sense.  There was a problem with the link tool.  I had added a text window so the link text could be changed from the filename as requested in #548.  For some reason that change caused the media links to be corrupted. It took a while to track it down.  So I've reverted to the link tool's original form and added your latest revision to mediamgr.js   It all seems to be working well, both in hogfather and stable-master:  https://github.com/turnermm/ckgedit/archive/hogfather_rc3a.zip.

**Comment by turnermm** [comment]
@PankajRasgania I have been using the version of mediamgr which checks for `valu == ""`.  This was entirely fortuitous, but it has been working without incident.  So I have been checking over the code to see exactly what is happening. 

First, the search box is always empty when the media manager starts up. Then whenever  a selection is made, it closes down; the next time it's called it restarts.    When the media manage is called, it is called from either the link dialog or the image dialog, each of which calls it with its own query string.  The query string is seen in the location bar at the top of the page.  It's by means of this query string that the decision is made as to whether this call is with respect to a link or an image.  So before any selection is made, the proper query string is then attached to the search form.  This is why when you use the link dialog and select an image, you get a link and not an image, and if you use the image dialog and select a non-image, the image dialog shows a broken image.  

It's taken me a while to work my way around this.  I didn't write the code myself.

**Comment by turnermm** [comment]
I guess they way to look at it, is that for each instance of the media manager in ckgedit, there is only one transaction, unlike the full dokuwiki media manager, which can remain open.

**Comment by PankajRasgania** [comment]
Yes. But we can live with that.

**Comment by PankajRasgania** [comment]
In the final commits you have used `window.load` function.  See [this](https://github.com/turnermm/ckgedit/issues/565#issuecomment-655967462)

**Comment by turnermm** [comment]
I tested your mediamgr.js onthree servers, all with greebo and hogfather.  On one server it failed in both dokuwiki distributions.  In the other, if failed in hogfather but succeeded in greebo.  Te third sever had similar results.  One server, with the one successful instance, is Centos, the others are Ubuntu.  They are all running php7, including the most recent, 7.4  

The javascript console for one of the failed instances showed a reference error, it could not find the jQuery needed for jQuery.ready() at the top of the media manager.  The version that succeeded in all cases used windows.onload.   From what I have read, Windows.onload loads last, after everything else, so it comes in after dokuwiki has finished with its deferred  javascript, and so jQuery will be available.  The point Andi makes in his discussions of hogfather is that  calls to the deferred objects have to wait until after the deferred code has been loaded.  Window.onload makes sure this will happen.  

You seem to have had a different experience from what is described above.

**Comment by PankajRasgania** [comment]
Yes I am having a different experience. I don't know why.  Once case where jQuery might fail is when mediamgr.js in not included in defer stack but jQuery library is included in it. In that case mediamgr.js will execute before jQuery library.

Also I am seeing a behaviour which I am not sure if it is intended by CKGEdit plugin.
When I install a new instance of Hogfather RC3, by default all scripts are defered and you can see the defer tag in scripts. But as soon as I install CKGEdit plugin using extension manager, the defer tag in scripts is not shown. The defer_js config has no effect when CKGEdit is enabled. I don't know what is happening. I observed this bevahiour on Windows and CentOs.

**Comment by turnermm** [comment]
> When I install a new instance of Hogfather RC3, by default all scripts are defered and you can see the defer tag in scripts

In earlier versions of ckgedit several scripts were output early.  But for hogfather, I refactored all script handling and moved them all to the meta sction of the HTML file. In following Andi's instruction all scripts  are now output with the defer tag except for ckeditor.js, which is the CKEditor's javascript.  That gets marked as deferred only if the ckgedit preload option was set. I did that because ckeditor.js is a higed file.  But thinking it over, I should probably remove that option in hogfather or at the least coordinate it with the defer_js option.  So the difference you are seeing is the difference in the handling of the scripts between hogfather and the stable-master branch, which is what is downloaded when using the extension manager.  

> Once case where jQuery might fail is when mediamgr.js in not included in defer stack but jQuery library is included in it. 
Now that is probably my fault.  mediamgr.js is loaded from mediamanager.php without the defer flag.  It slipped past me when refactoring the script output.  But I'm not sure why that should make a difference in this case.  Let me play around with it.

**Comment by turnermm** [comment]
I added the defer flag to mediamgr.js.   I'm currently at https://github.com/turnermm/ckgedit/archive/hogfather_rc3a.zip

**Comment by PankajRasgania** [comment]
Yes you are right. Actually I had doubt regarding when window.onload event is fired. Everywhere it says that is fired after resource loading. But if the resource is a script, does it get fired after script has been downloaded or after its execution. 

After some tests, I got the answer that it is fired after all of the defered script has executed. So no worries. [This article](https://flaviocopes.com/javascript-async-defer/#:~:text=Scripts%20marked%20defer%20are%20executed,domComplete%20event%2C%20and%20then%20onLoad%20.) explained it very well.

Also the op of #565 in this [comment](https://github.com/turnermm/ckgedit/issues/565#issuecomment-655916065) has some pointed out that something was not working. I wrongly diagnosed the problem as one with `window.onload`. Though the solution (using jQuery's ready) somehow worked for him. If you go to the url of mediamanager of his wiki - [https://senduwiki.org/lib/exe/mediamanager.php?ns=&edid=wiki__text&onselect=ckg_edit_mediaman_insertlink&ckg_media=link&CKEditor=wiki__text&CKEditorFuncNum=1&langCode=en](https://senduwiki.org/lib/exe/mediamanager.php?ns=&edid=wiki__text&onselect=ckg_edit_mediaman_insertlink&ckg_media=link&CKEditor=wiki__text&CKEditorFuncNum=1&langCode=en) you see that action attribute of the search form is not getting updated. It seems that the code inside the `window.onload` function is not getting executed. May be the whole script is not getting executed. I don't know why it is happening. But this should not happen.

**Comment by turnermm** [comment]
I had seen but not really read that article.  I did read it this tiime, and it is good. Dokuwiki loads all of the php files when going into edit mode. That means that mediamanager.php is called only once, when the editor starts, and with it mediamgr.js.  I wonder if it makes a difference in which order the files are stored in the action directory.

**Comment by turnermm** [comment]
> But if the resource is a script, does it get fired after script has been downloaded or after its execution.

I had the same question.
